Source/WebCore/rendering/RenderThemeIOS.mm

```diff
@@ -1570,7 +1570,11 @@ static bool shouldUseConvexGradient(const Color& backgroundColor)
     ASSERT(!forVisitedLink);
 
     auto& cache = colorCache(options);
-    return cache.systemStyleColors.ensure(cssValueID, [this, cssValueID, options] () -> Color {
+    auto it = cache.systemStyleColors.find(cssValueID);
+    if (it != cache.systemStyleColors.end())
```

Comment:
Can you explain why this change is necessary? I'm not sure I see the difference between this and `ensure`.

Reply:
This change prevents entering `cache.system StyleColors.ensure()` method twice.

This can happen because we call `RenderTheme::systemColor()` from `cache.systemStyleColors.ensure()`  when `globalCSSValueToSystemColorMap()` doesn't contain the color we're trying to resolve:

https://github.com/WebKit/WebKit/blob/b38121da42a0483fdb1bc13782f1c7a8059de100/Source/WebCore/rendering/RenderThemeIOS.mm#L1573-L1586

For deprecated colors (e.g. `CSSValueThreedface`) `RenderTheme::systemColor()` will call `RenderThemeIOS::systemColor()` again with "same-as" color (e.g. `CSSValueButtonface`).
https://github.com/WebKit/WebKit/blob/cb9b5c3f9215562642c1d48d92552c2586fdf474/Source/WebCore/rendering/RenderTheme.cpp#L1898-L1899

If this color is not yet cached we'll end up in `enusre()` again. This is "modifying-while-modifying" kind of problem. Which can lead to some really weird behavior.

Reply:
Thanks!